FreeRTOS_Sockets.c

```diff
@@ -3647,155 +3935,214 @@ void vSocketWakeUpUser( FreeRTOS_Socket_t * pxSocket )
+        }
+
+        return xByteCount;
+    }
+#endif /* ipconfigUSE_TCP */
+/*-----------------------------------------------------------*/
+
+#if ( ipconfigUSE_TCP == 1 )
+
+/**
+ * @brief After FreeRTOS_recv() has checked the validity of the parameters,
+ *        this routine will wait for data to arrive in the stream buffer.
+ *
+ * @param[in] pxSocket: The socket owning the connection.
+ * @param[out] pxEventBits: A bit-mask of socket events will be set:
+ *             eSOCKET_RECEIVE, eSOCKET_CLOSED, and or eSOCKET_INTR.
+ * @param[in] xFlags: the flags that were passed to FreeRTOS_recv().
+ */
+    static BaseType_t prvRecvWait( FreeRTOS_Socket_t * pxSocket,
+                                   EventBits_t * pxEventBits,
+                                   BaseType_t xFlags )
+    {
+        BaseType_t xByteCount = 0;
         TickType_t xRemainingTime;
         BaseType_t xTimed = pdFALSE;
         TimeOut_t xTimeOut;
-        EventBits_t xEventBits = ( EventBits_t ) 0;
+        EventBits_t xEventBits = ( EventBits_t ) 0U;
 
-        /* Check if the socket is valid, has type TCP and if it is bound to a
-         * port. */
-        if( prvValidSocket( pxSocket, FREERTOS_IPPROTO_TCP, pdTRUE ) == pdFALSE )
-        {
-            xByteCount = -pdFREERTOS_ERRNO_EINVAL;
-        }
-        else if( ( ( ( uint32_t ) xFlags & ( uint32_t ) FREERTOS_ZERO_COPY ) != 0U ) &&
-                 ( pvBuffer == NULL ) )
+        if( pxSocket->u.xTCP.rxStream != NULL )
         {
-            /* In zero-copy mode, pvBuffer is a pointer to a pointer ( not NULL ). */
-            xByteCount = -pdFREERTOS_ERRNO_EINVAL;
+            xByteCount = ( BaseType_t ) uxStreamBufferGetSize( pxSocket->u.xTCP.rxStream );
         }
-        else
+
+        while( xByteCount == 0 )
         {
-            if( pxSocket->u.xTCP.rxStream != NULL )
-            {
-                xByteCount = ( BaseType_t ) uxStreamBufferGetSize( pxSocket->u.xTCP.rxStream );
-            }
-            else
+            eIPTCPState_t eType = ( eIPTCPState_t ) pxSocket->u.xTCP.ucTCPState;
+
+            if( ( eType == eCLOSED ) ||
+                ( eType == eCLOSE_WAIT ) || /* (server + client) waiting for a connection termination request from the local user. */
+                ( eType == eCLOSING ) )     /* (server + client) waiting for a connection termination request acknowledgement from the remote TCP. */
             {
-                xByteCount = 0;
+                /* Return -ENOTCONN, unless there was a malloc failure. */
+                xByteCount = -pdFREERTOS_ERRNO_ENOTCONN;
+
+                if( pxSocket->u.xTCP.bits.bMallocError != pdFALSE_UNSIGNED )
+                {
+                    /* The no-memory error has priority above the non-connected error.
+                     * Both are fatal and will lead to closing the socket. */
+                    xByteCount = -pdFREERTOS_ERRNO_ENOMEM;
+                }
+
+                break;
```

Comment:
This usage does not fall under the MISRA Rule 15.4 exception.

Reply:
MISRA C 2012 Rule 15.4 seems to be advisory. See my other comments in [PR #236](https://github.com/FreeRTOS/FreeRTOS-Plus-TCP/pull/236#discussion_r636754135)